src/utils/dlck/dlck_engine.c

```diff
@@ -0,0 +1,640 @@
+
+	D_ASSERT(tls != NULL);
+
+	if (bio_nvme_configured(SMD_DEV_TYPE_META)) {
+		rc = ABT_eventual_set(xs->nvme_poll_done, NULL, 0);
+		rc = dss_abterr2der(rc);
+		if (rc != DER_SUCCESS) {
+			goto fail;
+		}
+
+		rc = ABT_thread_join(xs->nvme_poll.thread);
+		rc = dss_abterr2der(rc);
+		if (rc != DER_SUCCESS) {
+			goto fail;
+		}
+
+		rc = ABT_thread_free(&xs->nvme_poll.thread);
+		rc = dss_abterr2der(rc);
+		if (rc != DER_SUCCESS) {
+			/**
+			 * After the NVMe polling thread joined we can safely free TLS irrespective
+			 * of the error occurred while freeing the thread.
+			 */
+		}
+
+		dmi = dss_get_module_info();
+		D_ASSERT(dmi != NULL);
+		bio_xsctxt_free(dmi->dmi_nvme_ctxt);
+	}
+
+	dss_tls_fini(tls);
+
+fail:
+	/**
+	 * In case of a fail we can't join/free the NVMe polling thread nor free TLS which may
+	 * result in a SIGSEGV. The best we can do is to leave the resources as they are and pass
+	 * error the caller.
+	 */
+
+	return rc;
+}
+
+static void
+dlck_engine_xstream_fini_ult(void *arg)
+{
+	struct dlck_xstream *xs = arg;
+
+	xs->ult_rc = dlck_engine_xstream_fini(xs);
+}
+
+/**
+ * Create and initialize daos_sys_0 execution stream (XS) and create all daos_io_* XSes.
+ * No daos_io_* initialization here yet. They ought to be initialized by the first ULT run in them.
+ *
+ * \param[in,out]	engine	Engine to start its XSes.
+ *
+ * \retval DER_SUCCESS	Success.
+ * \retval -DER_*	Error.
+ */
+static int
+xstream_start_all(struct dlck_engine *engine)
+{
+	struct dlck_xstream *xs;
+	struct dlck_ult      daos_sys_init;
+	int                  rc;
+
+	/** create and initialize daos_sys_0 execution stream (XS) */
+	xs         = &engine->xss[engine->targets]; /** there is one more XS than targets */
+	xs->tgt_id = -1;
+	rc         = dlck_xstream_create(xs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = dlck_ult_create(xs->pool, dlck_engine_xstream_init_ult, xs, &daos_sys_init);
+	if (rc != DER_SUCCESS) {
+		/** ULT has not been created - the daos_sys_0 XS can be safely freed */
+		(void)dlck_xstream_free(xs);
+		return dss_abterr2der(rc);
+	}
+
+	/** wait for the daos_sys_0 initialization to conclude */
+	rc = ABT_thread_join(daos_sys_init.thread);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("ULT has not joined - cannot safely free the daos_sys_0 XS\n");
+		return dss_abterr2der(rc);
+	}
+
+	rc = ABT_thread_free(&daos_sys_init.thread);
+	if (rc != ABT_SUCCESS) {
+		/** ULT has joined - the daos_sys_0 XS can be safely freed */
+		(void)dlck_xstream_free(xs);
+		return dss_abterr2der(rc);
+	}
+
+	if (xs->ult_rc != DER_SUCCESS) {
+		/** ULT has joined - the daos_sys_0 XS can be safely freed */
+		(void)dlck_xstream_free(xs);
+		return xs->ult_rc;
+	}
+
+	/**
+	 * The daos_sys_0 XS initialization succeeded. It may have spawned a NVMe polling ULT.
+	 */
+
+	/** create all daos_io_* execution streams (XS) */
+	for (int i = 0; i < engine->targets; ++i) {
+		xs         = &engine->xss[i];
+		xs->tgt_id = i;
+		rc         = dlck_xstream_create(xs);
```

Comment:
I didn't see how nvme polling ULTs are created for VOS xstreams

Reply:
A procedure executed as ULT on daos_io_* has to explicitly call `dlck_engine_xstream_init()`. It will start the NVMe polling ULT if necessary.

Reply:
Which procedure? I didn't see it in this PR?

Reply:
It depends on the user of the `dlck_engine*()` interface. In this PR is only one: dlck_test_helper. You can see how it is used here: https://github.com/daos-stack/daos/blob/janekmi/DAOS-17569-DLCK-DAE-records-recover/src/utils/dlck/tests/dlck_test_helper.c#L451

Reply:
Because there is only one request. Please see the diagrams here: https://github.com/daos-stack/daos/pull/16550#issuecomment-3243351555